src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,475 @@
+-/
+structure RefStreamEntry (α : Type) [DecidableEq α] [Hashable α] [DecidableEq α] (length : Nat) where
+  aig : AIG α
+  stream : RefStream aig length
+
+structure ShiftTarget (aig : AIG α) (w : Nat) where
+  stream : AIG.RefStream aig w
+  distance : Nat
+
+structure ArbitraryShiftTarget (aig : AIG α) (m : Nat) where
+  n : Nat
+  target : AIG.RefStream aig m
+  distance : AIG.RefStream aig n
+
+structure ExtendTarget (aig : AIG α) (newWidth : Nat) where
+  w : Nat
+  stream : AIG.RefStream aig w
+
+/--
+Evaluate an `AIG.Entrypoint` using some assignment for atoms.
+-/
+def denote (assign : α → Bool) (entry : Entrypoint α)  : Bool :=
+  go entry.ref.gate entry.aig.decls assign entry.ref.hgate entry.aig.inv
+where
+  go (x : Nat) (decls : Array (Decl α)) (assign : α → Bool) (h1 : x < decls.size)
+      (h2 : IsDag α decls) :
+      Bool :=
+    match h3 : decls[x] with
+    | .const b => b
+    | .atom v => assign v
+    | .gate lhs rhs linv rinv =>
+      have := h2 x _ _ _ _ h1 h3
+      let lval := go lhs decls assign (by omega) h2
+      let rval := go rhs decls assign (by omega) h2
+      xor lval linv && xor rval rinv
+
+scoped syntax "⟦" term ", " term "⟧" : term
+scoped syntax "⟦" term ", " term ", " term "⟧" : term
+
+macro_rules
+| `(⟦$entry, $assign⟧) => `(denote $assign $entry )
+| `(⟦$aig, $ref, $assign⟧) => `(denote $assign (Entrypoint.mk $aig $ref))
+
+@[app_unexpander AIG.denote]
+def unexpandDenote : Lean.PrettyPrinter.Unexpander
+  | `($(_) {aig := $aig, start := $start, inv := $hbound} $assign) =>
+    `(⟦$aig, ⟨$start, $hbound⟩, $assign⟧)
+  | `($(_) $entry $assign) => `(⟦$entry, $assign⟧)
+  | _ => throw ()
+
+def UnsatAt (aig : AIG α) (start : Nat) (h : start < aig.decls.size) : Prop :=
+  ∀ assign, ⟦aig, ⟨start, h⟩, assign⟧ = false
+
+def Entrypoint.Unsat (entry : Entrypoint α) : Prop :=
+  entry.aig.UnsatAt entry.ref.gate entry.ref.hgate
+
+/--
+An input to an AIG gate.
+-/
+structure Fanin (aig : AIG α) where
+  /--
+  The node we are referring to.
+  -/
+  ref : Ref aig
+  /--
+  Whether the node is inverted
+  -/
+  inv : Bool
+
+@[inline]
+def Fanin.cast {aig1 aig2 : AIG α} (fanin : Fanin aig1)
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    Fanin aig2 :=
+  { fanin with ref := fanin.ref.cast h }
+
+/--
+The input type for creating AIG and gates from scratch.
+-/
+structure GateInput (aig : AIG α) where
+  lhs : Fanin aig
+  rhs : Fanin aig
+
+@[inline]
+def GateInput.cast {aig1 aig2 : AIG α} (input : GateInput aig1)
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    GateInput aig2 :=
+  { input with lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+/--
+Build an AIG gate in `aig`. Note that this version is only meant for proving,
+for production purposes use `AIG.mkGateCached` and equality theorems to this one.
+-/
+def mkGate (aig : AIG α) (input : GateInput aig) : Entrypoint α :=
+  let lhs := input.lhs
+  let rhs := input.rhs
+  let g := aig.decls.size
```

Comment:
This applies to many functions, but why do you have so many `let`s for things that are then only used once downstream? I don't think it really improves readability much.

Reply:
In this particular function there is no precise reason. In the functions that work on AIGs see my remarks about shadowing etc.

Reply:
Makes sense, thanks for the explanation.